// Effect.Class.Console/index.js
"use strict";
var Control_Semigroupoid = require("../Control.Semigroupoid/index.js");
var Data_Function = require("../Data.Function/index.js");
var Data_Show = require("../Data.Show/index.js");
var Data_Unit = require("../Data.Unit/index.js");
var Effect_Class = require("../Effect.Class/index.js");
var Effect_Console = require("../Effect.Console/index.js");
var warnShow = function (dictMonadEffect) {
    return function (dictShow) {
        return function ($12) {
            return Effect_Class.liftEffect(dictMonadEffect)(Effect_Console.warnShow(dictShow)($12));
        };
    };
};
var warn = function (dictMonadEffect) {
    return function ($13) {
        return Effect_Class.liftEffect(dictMonadEffect)(Effect_Console.warn($13));
    };
};
var logShow = function (dictMonadEffect) {
    return function (dictShow) {
        return function ($14) {
            return Effect_Class.liftEffect(dictMonadEffect)(Effect_Console.logShow(dictShow)($14));
        };
    };
};
var log = function (dictMonadEffect) {
    return function ($15) {
        return Effect_Class.liftEffect(dictMonadEffect)(Effect_Console.log($15));
    };
};
var infoShow = function (dictMonadEffect) {
    return function (dictShow) {
        return function ($16) {
            return Effect_Class.liftEffect(dictMonadEffect)(Effect_Console.infoShow(dictShow)($16));
        };
    };
};
var info = function (dictMonadEffect) {
    return function ($17) {
        return Effect_Class.liftEffect(dictMonadEffect)(Effect_Console.info($17));
    };
};
var errorShow = function (dictMonadEffect) {
    return function (dictShow) {
        return function ($18) {
            return Effect_Class.liftEffect(dictMonadEffect)(Effect_Console.errorShow(dictShow)($18));
        };
    };
};
var error = function (dictMonadEffect) {
    return function ($19) {
        return Effect_Class.liftEffect(dictMonadEffect)(Effect_Console.error($19));
    };
};
module.exports = {
    log: log,
    logShow: logShow,
    warn: warn,
    warnShow: warnShow,
    error: error,
    errorShow: errorShow,
    info: info,
    infoShow: infoShow
};
